### PR TITLE
Enable CircleCI Test Insights

### PR DESCRIPTION
**Pull request by chrzaszcz**
The goal is to enable CircleCI Test Insights for big tests, that can automatically detect flaky and most failing tests.

This can be achieved by using the `cth_surefire` common test hook.
However, the hook contained a few bugs, so:

1. `cth_surefire` is copied from Erlang/OTP to big tests.
2. The fixes are applied.

When reviewing, **check the last commit** to see the diff in `cth_surefire`.

Here you can see the Insights in action:
- [Reports with some failures for previous runs](https://app.circleci.com/pipelines/github/esl/MongooseIM?branch=circleci-test-insights)
- [Insights with flaky tests](https://app.circleci.com/insights/github/esl/MongooseIM/workflows/build_and_test/tests?branch=circleci-test-insights)

Screenshot of a flaky test from reports:

### CODE REVIEW

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/157000) / small_tests / 9b04c9e6f41bff4779d811c134929c181ce16916
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3899/157000/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3899/157000/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-12-09_14.57.13/lib.mongooseim.logs/run.2022-12-09_14.57.13/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/156991) / small_tests / df88afd97a643a8b6d83d53153b6340950cda9bd
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3899/156991/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3899/156991/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-12-09_14.56.25/lib.mongooseim.logs/run.2022-12-09_14.56.25/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/157037) / small_tests / de2a98026690206e4bbab502d2dca9d858572e47
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3899/157037/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3899/157037/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-12-09_15.24.09/lib.mongooseim.logs/run.2022-12-09_15.24.09/suite.log.html)

**Comment by mongoose-im** [comment]
[ldap_mnesia_24](https://circleci.com/gh/esl/MongooseIM/157066) / ldap_mnesia / 294b79181c710e02089d36b2ece69f9dd6d0bd16
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3899/157066/ldap_mnesia.24.3.4)

**Comment by mongoose-im** [comment]
[ldap_mnesia_24](https://circleci.com/gh/esl/MongooseIM/157081) / ldap_mnesia / bcdc6fb29679edd3d8a26d6407f3d26cd0c06426
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3899/157081/ldap_mnesia.24.3.4)

**Comment by mongoose-im** [comment]
[pgsql_mnesia_24](https://circleci.com/gh/esl/MongooseIM/157075) / pgsql_mnesia / bcdc6fb29679edd3d8a26d6407f3d26cd0c06426
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3899/157075/pgsql_mnesia.24.3.4)

**Comment by mongoose-im** [comment]
[dynamic_domains_pgsql_mnesia_24](https://circleci.com/gh/esl/MongooseIM/157097) / pgsql_mnesia / aa4e3e48ed507bcf12196cdd4d57aee02254f173
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3899/157097/pgsql_mnesia.24.3.4)

**Comment by mongoose-im** [comment]
[dynamic_domains_pgsql_mnesia_24](https://circleci.com/gh/esl/MongooseIM/157135) / pgsql_mnesia / 2b6925a91fc097168d2fbd59b15ba73702c83b90
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3899/157135/pgsql_mnesia.24.3.4)

**Comment by mongoose-im** [comment]
[dynamic_domains_pgsql_mnesia_24](https://circleci.com/gh/esl/MongooseIM/157163) / pgsql_mnesia / 7be0dcaa02f5b6c76a143287c565bd34aa163a29
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3899/157163/pgsql_mnesia.24.3.4)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/157183) / small_tests / 2a1c1aac708864843089b86c5bd8c530d5d22706
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3899/157183/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3899/157183/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-12-09_17.28.10/lib.mongooseim.logs/run.2022-12-09_17.28.11/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/157238) / small_tests / 5ec1fc5e1c4b67cbd2601cd5b4ac029a2ec73026
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3899/157238/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3899/157238/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-12-12_08.15.46/lib.mongooseim.logs/run.2022-12-12_08.15.46/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/157853) / small_tests / 9fbe8df8c602d76c5fb0df73edbf7f7335b16258
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3899/157853/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3899/157853/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-12-15_07.25.56/lib.mongooseim.logs/run.2022-12-15_07.25.56/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/158091) / small_tests / 9108acb35ebdfa377b9889feaa20b91d677398fb
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3899/158091/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3899/158091/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-12-15_13.17.37/lib.mongooseim.logs/run.2022-12-15_13.17.38/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/158501) / small_tests / 341870137f38128fdc061ad471a202988110628d
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3899/158501/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3899/158501/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-12-20_08.19.10/lib.mongooseim.logs/run.2022-12-20_08.19.10/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/158549) / small_tests / edd358072df63348a1e92bd83b662f2b7b6adefc
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3899/158549/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3899/158549/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-12-20_15.07.45/lib.mongooseim.logs/run.2022-12-20_15.07.46/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/158623) / small_tests / 07d220c88d09f263cb4f00e9df342d530f2e3d4c
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3899/158623/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3899/158623/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-12-20_16.05.55/lib.mongooseim.logs/run.2022-12-20_16.05.56/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/158651) / small_tests / 1850e4602ce2f4c031d2dac73ab7f05e16436fe3
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3899/158651/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3899/158651/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-12-20_17.14.54/lib.mongooseim.logs/run.2022-12-20_17.14.55/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/158684) / small_tests / f9a384045f2795d4de4bfc9ff3b55033d5b29b2f
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3899/158684/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3899/158684/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-12-21_08.11.06/lib.mongooseim.logs/run.2022-12-21_08.11.06/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/158711) / small_tests / 199f1b4fb753b968158f69f4a4e493aa4f152c7a
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3899/158711/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3899/158711/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-12-21_08.47.07/lib.mongooseim.logs/run.2022-12-21_08.47.07/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/158757) / small_tests / 046fe2892db0621ea24ca65adba82b782a3eafa0
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3899/158757/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3899/158757/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-12-21_09.43.18/lib.mongooseim.logs/run.2022-12-21_09.43.19/suite.log.html)

File: big_tests/Makefile [hunk 1]
```diff
@@ -4,6 +4,7 @@ all: test
 # user overridable
 TESTSPEC ?= default.spec
 PRESET   ?= all
+CT_HOOKS ?= ""
```
Comment (kamilwaz) [inline big_tests/Makefile:7]: This one isn't needed. If we prefer to keep it to remind us about the existence of the variable, then I would change it to the following:

```suggestion
CT_HOOKS ?=
```

**Review by kamilwaz — CHANGES_REQUESTED** [review]
Looks good :+1:, just one small comment.

File: big_tests/Makefile (hunk 1 above)
Comment (chrzaszcz) [reply big_tests/Makefile:7]: The intention is to have a list of user-overridable variables, see the comment in line 4.
If you think that `""` looks verbose, I can remove it, it makes no big difference to be honest.

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/158835) / small_tests / 651bdd5d0de7895494f6183bca15ab86e63d50a0
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3899/158835/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3899/158835/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-12-21_13.08.57/lib.mongooseim.logs/run.2022-12-21_13.08.58/suite.log.html)

**Review by kamilwaz — APPROVED** [review]

**Pull request merged**